File: src/main/java/me/devsnox/battlekits/KitManager.java
package me.devsnox.battlekits;

import de.tr7zw.itemnbtapi.NBTItem;
import me.devsnox.battlekits.kits.kit.BattleKit;
import me.devsnox.battlekits.kits.entites.KitPlayer;
import me.devsnox.battlekits.kits.kit.KitType;
import me.devsnox.battlekits.kits.loader.BasicPlayerLoader;
import me.devsnox.battlekits.listeners.FrameListener;
import me.devsnox.battlekits.kits.loader.BasicKitLoader;
import me.devsnox.battlekits.kits.loader.PlayerLoader;
import me.devsnox.battlekits.utils.ItemBuilder;
import me.devsnox.itemprotection.api.ItemProtectionAPI;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.Plugin;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class KitManager {

    private Plugin plugin;

    private BasicKitLoader kitLoader;

    private PlayerLoader playerLoader;
    private Map<UUID, KitPlayer> playerCache;
    private ArrayList<Player> openingChache;

    private FrameListener frameListener;

    public KitManager(Plugin plugin) {
        this.plugin = plugin;
    }

    public void enable() {
        this.playerLoader = new BasicPlayerLoader(this, plugin.getDataFolder());
        this.playerCache = new HashMap<>();
        this.openingChache = new ArrayList<>();
        this.kitLoader = new BasicKitLoader(this.plugin.getDataFolder());
        kitLoader.loadKits();

        for(Player player : Bukkit.getOnlinePlayers()) {
            this.loadPlayer(player.getUniqueId());
        }
    }

    public void disable() {
        for(Player player : Bukkit.getOnlinePlayers()) {
            this.savePlayer(player.getUniqueId());
            player.closeInventory();
        }
        this.kitLoader.saveKits();
    }

    public Plugin getPlugin() {
        return plugin;
    }

    public KitPlayer getPlayer(UUID uuid) {
        return this.playerCache.get(uuid);
    }

    public void loadPlayer(UUID uuid) {
        this.playerCache.put(uuid, this.playerLoader.loadPlayer(uuid));
    }

    public void savePlayer(UUID uuid) {
        this.playerLoader.savePlayer(this.playerCache.get(uuid));
        this.playerCache.remove(uuid);
    }

    public Map<Integer, BattleKit> getKits() {
        return this.kitLoader.getKits();
    }


    public FrameListener getFrameListener() {
        return frameListener;
    }

    public void setFrameListener(FrameListener frameListener) {
        this.frameListener = frameListener;
    }

    public void updateKit(BattleKit battleKit) {
        this.kitLoader.updateKit(battleKit);
    }

    public ArrayList<BattleKit> getKitsByType(KitType kitType) {
        ArrayList<BattleKit> kits = new ArrayList<>();

        for(BattleKit battleKit : getKits().values()) {
            if(battleKit.getKitType() == kitType) {
                kits.add(battleKit);
            }
        }

        return kits;
    }

    public ItemStack getShard(int id) {
        BattleKit battleKit = this.getKits().get(id);

        ItemBuilder itemBuilder = new ItemBuilder();

        itemBuilder.setType(Material.PRISMARINE_CRYSTALS);
        itemBuilder.setGlow(true);

        itemBuilder.setDisplayName("§b§lShard §8- " + battleKit.getKitType().getColor() + battleKit.getName());

        NBTItem nbtItem = new NBTItem(itemBuilder.build());

        nbtItem.setInteger("kitId", id);
        nbtItem.setDouble("chance", battleKit.getKitType().getChance());

        return ItemProtectionAPI.getInstance().addProtection(nbtItem.getItem());
    }

    public ItemStack getBooster(double percentage) {
        ItemBuilder itemBuilder = new ItemBuilder();

        itemBuilder.setType(Material.GOLD_NUGGET);
        itemBuilder.setGlow(true);

        itemBuilder.setDisplayName("§6§lBooster §8- §7+ §b" + percentage + "§7%");

        NBTItem nbtItem = new NBTItem(itemBuilder.build());

        nbtItem.setDouble("boost", percentage);

        return nbtItem.getItem();
    }

    public void setPlayerOpening(Player player, boolean active) {
        if(active == true) {
            this.openingChache.add(player);
        } else {
            this.openingChache.remove(player);
        }
    }

    public boolean isOpening(Player player) {
        return this.openingChache.contains(player);
    }
}

File: spigot/src/main/kotlin/me/devsnox/battlekits/KitManager.kt
package me.devsnox.battlekits

import de.tr7zw.itemnbtapi.NBTItem
import me.devsnox.battlekits.functions.KitsB
import me.devsnox.battlekits.functions.OpeningChache
import me.devsnox.battlekits.functions.PlayerCache
import me.devsnox.battlekits.kits.entites.KitPlayer
import me.devsnox.battlekits.kits.kit.BattleKit
import me.devsnox.battlekits.kits.kit.KitType
import me.devsnox.battlekits.kits.loader.KitLoader
import me.devsnox.battlekits.kits.loader.PlayerLoader
import me.devsnox.itemprotection.api.ItemProtectionFactory
import net.darkdevelopers.darkbedrock.darkness.spigot.builder.item.ItemBuilder
import net.darkdevelopers.darkbedrock.darkness.spigot.functions.glow
import net.darkdevelopers.darkbedrock.darkness.spigot.utils.Utils
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.plugin.Plugin
import java.util.*

class KitManager(
    val plugin: Plugin,
    private val players: Collection<Player> = Utils.players
) {

    private var playerCache: PlayerCache = mutableMapOf()
    private var openingChache: OpeningChache = mutableListOf()
    private var playerLoader: PlayerLoader = PlayerLoader(this)
    private var kitLoader: KitLoader = KitLoader(plugin)

    val kits: KitsB get() = kitLoader.kits

    suspend fun enable() {
        kitLoader.loadKits()

        for (player in players) loadPlayer(player.uniqueId)
    }

    suspend fun disable() {
        for (player in players) {
            savePlayer(player.uniqueId)
            player.closeInventory()
        }
        kitLoader.saveKits()
    }

    fun getPlayer(uuid: UUID): KitPlayer? = playerCache[uuid]

    suspend fun loadPlayer(uuid: UUID) {
        playerCache[uuid] = playerLoader.loadPlayer(uuid)
    }

    suspend fun savePlayer(uuid: UUID) {
        val kitPlayer = playerCache[uuid] ?: return
        playerLoader.savePlayer(kitPlayer)
        playerCache.remove(uuid)
    }

    fun updateKit(battleKit: BattleKit): Unit = kitLoader.updateKit(battleKit)

    fun getKitsByType(kitType: KitType): List<BattleKit> = kits.values.filter { it.kitType == kitType }

    fun getShard(id: Int): ItemStack {
        val battleKit = kits[id] ?: return ItemStack(Material.BARRIER)
        val itemBuilder = ItemBuilder(Material.PRISMARINE_CRYSTALS).glow()
            .setDisplayName("§b§lShard §8- ${battleKit.kitType.color}${battleKit.name}")

        val nbtItem = NBTItem(itemBuilder.build())
        nbtItem.setInteger("kitId", id)
        nbtItem.setDouble("chance", battleKit.kitType.chance)
        return ItemProtectionFactory.itemProtection!!.addProtection(nbtItem.item)
    }

    fun getBooster(percentage: Double): ItemStack {
        val itemBuilder = ItemBuilder(Material.GOLD_NUGGET).glow()
            .setDisplayName("§6§lBooster §8- §7+ §b$percentage§7%")
        return NBTItem(itemBuilder.build()).apply { setDouble("boost", percentage) }.item
    }

    fun setPlayerOpening(player: Player, active: Boolean) {
        if (active) openingChache.add(player) else openingChache.remove(player)
    }

    fun isOpening(player: Player): Boolean = player in openingChache
}
